### DISCUSSION TITLE
Hard crash mid-map on std, probably related to texture loading failure.

### DISCUSSION THREAD

**Discussion by Syndace** [ppy/osu#16388] [Q&A]
The game crashed on me during std gameplay. On the last frame that remained on screen for like half a second before disappearing, I noticed a small-ish black box where some hit object was probably supposed to appear. I guess the tail of `runtime.log` gives a good hint about what went wrong:

```
[...]
2022-01-09 14:53:19 [verbose]: 📺 OsuScreenStack#401(depth:5) resume to PlaySongSelect#481
2022-01-09 14:53:20 [verbose]: updating selection with beatmap:5829 ruleset:0
2022-01-09 14:53:20 [verbose]: beatmap changed from "inabakumori - Lost Umbrella (Ryuusei Aika) [Jounzan's PTSD Extra]" to "Yunomi - Mentai Cosmic (alacat) [Extra]"
2022-01-09 14:53:20 [verbose]: Game-wide working beatmap updated to Yunomi - Mentai Cosmic (alacat) [Extra]
2022-01-09 14:53:20 [verbose]: Song select working beatmap updated to Yunomi - Mentai Cosmic (alacat) [Extra]
2022-01-09 14:53:21 [verbose]: Texture could not be loaded via STB; falling back to ImageSharp: STBI could not load an image from the provided data: 畏⁴景洠浥牯y
2022-01-09 14:53:21 [verbose]: Texture could not be loaded via STB; falling back to ImageSharp: STBI could not load an image from the provided data: 畏⁴景洠浥牯y
2022-01-09 14:53:21 [verbose]: Texture could not be loaded via STB; falling back to ImageSharp: STBI could not load an image from the provided data: 畏⁴景洠浥牯y
2022-01-09 14:53:23 [verbose]: 📺 OsuScreenStack#401(depth:5) suspended PlaySongSelect#481 (waiting on PlayerLoader#163)
2022-01-09 14:53:23 [verbose]: 📺 OsuScreenStack#401(depth:6) loading PlayerLoader#163
2022-01-09 14:53:23 [verbose]: 📺 OsuScreenStack#401(depth:6) entered PlayerLoader#163
2022-01-09 14:53:25 [verbose]: 📺 OsuScreenStack#401(depth:6) suspended PlayerLoader#163 (waiting on SoloPlayer#204)
2022-01-09 14:53:25 [verbose]: 📺 OsuScreenStack#401(depth:7) entered SoloPlayer#204
(end of file)
```

I could not reproduce the crash. After restarting the game, I selected the same beatmap and played through it just fine; no `Texture could not be loaded [...]` lines in the log either. However, I think I've encountered the same crash on a different beatmap before (on an earlier patch, maybe like two or three weeks ago). Sadly back then I wasn't smart enough to make a copy of the logs before restarting the game, thus the logs of that crash are lost.

Playing on Windows 10, running the most recent version of lazer.

The other logs look rather uninteresting, however I'll keep them around in case you need them anyway.

(Great game, even greater by going open-source. Thank you <3)

**Comment by bdach** [comment]
please attach the full logs as well as link the map.

is there a chance your system memory was low at the time or something of that sort?

**Reply by Syndace** [reply]
Alright here you go:

[database.log](https://github.com/ppy/osu/files/7835405/database.log)
[network.log](https://github.com/ppy/osu/files/7835406/network.log)
[performance.log](https://github.com/ppy/osu/files/7835407/performance.log)
[runtime.log](https://github.com/ppy/osu/files/7835408/runtime.log)
[updater.log](https://github.com/ppy/osu/files/7835409/updater.log)

Well I don't actively monitor my ram usage usually. 16gb are available to Windows and lazer was the only application running, so there should have been around 12gb available to lazer. I reckon lazer usually doesn't require 12+gigs of ram, but I guess there's always the possibility of a (severe) memory leak. It was a rather long session so it's not impossible that the memory was slowly filled up by some leak.

Maybe you could increase the verbosity of the `Texture could not be loaded via STB` log line just a tad to include a reason like "low memory" :D

[map link](https://osu.ppy.sh/beatmapsets/436773#osu/940706)

**Reply by Syndace** [reply]
Is there anything I can do to make sure that I'll produce more useful information if I ever encounter the same crash? Like running a debug build with a profiler attached or anything of that sort? Otherwise, if memory shortage is your best guess, I'll just keep an eye on memory usage in the future.

**Reply by Syndace** [reply]
Another crash!

[database.log](https://github.com/ppy/osu/files/7839563/database.log)
[network.log](https://github.com/ppy/osu/files/7839564/network.log)
[performance.log](https://github.com/ppy/osu/files/7839565/performance.log)
[runtime.log](https://github.com/ppy/osu/files/7839566/runtime.log)
[updater.log](https://github.com/ppy/osu/files/7839567/updater.log)

Slightly different flavor than the last one but you were correct about the cause: before crashing, lazer used a whopping 8.5gb of memory. That makes for 80% of system memory used overall. There would have been another two gigabytes for lazer to swallow but it seems like it decided to crash before using those too. By professionally tracing the memory footprint (i.e. keeping an eye on the task manager lol), I found that:
- gameplay does not seem to significantly increase memory usage. The usage goes up slowly and then suddenly down by a rather large chunk (probably just a visualization of garbage collection at work).
- navigating the library/switching between collections/searching songs etc. does seem to increase memory usage by a lot at times, especially browsing the public beatmap listing for a while easily adds a gigabyte to the memory used (dunno how much of that is expected due to caching and stuff)
- the most obvious problem: leaving the game idle in song select for around 1:40 hours added a whopping three gigs to the memory used. In the last ~150 lines of `runtime.log` you can see that texture loading issues set in shortly after the break and the crash follows soon after.

I know devs _love_ it when users who don't know anything about the software wildly guess about bugs, so I'm gonna do exactly that (<3): The game probably doesn't do a whole lot while idle in song select other than polling chat and looping the song of the currently selected map, my money is on that song-looping having a leak.

Let me know if there's a way I can produce more useful information for you.

**Reply by bdach** [reply]
rather than guessing at random, it would be better if you captured a process dump via task manager (right click process -> "create dump file") when memory usage is overly high and sent it over to pe@ppy.sh. do NOT post that publically as it may contain sensitive data.

however, on checking the log, there are a few suspicious threads:

* does disabling hardware video rendering change anything?
* you have some aspire maps imported (like XNOR XNOR XNOR) that may be inadvertently causing this. is there a visible memory usage spike after such beatmaps are selected?
